src/hotspot/share/classfile/placeholders.cpp

```diff
@@ -116,11 +256,30 @@ Symbol* PlaceholderTable::find_entry(unsigned int hash,
   return (probe != NULL ? probe->klassname() : NULL);
 }
 
-  // find_and_add returns probe pointer - old or new
-  // If no entry exists, add a placeholder entry
-  // If entry exists, reuse entry
-  // For both, push SeenThread for classloadAction
-  // If LOAD_SUPER, this is used for circularity detection for instanceklass loading.
+static const char* action_to_string(PlaceholderTable::classloadAction action) {
+  switch (action) {
+  case PlaceholderTable::LOAD_INSTANCE: return "LOAD_INSTANCE";
+  case PlaceholderTable::LOAD_SUPER:    return "LOAD_SUPER";
+  case PlaceholderTable::DEFINE_CLASS:  return "DEFINE_CLASS";
+ }
+ return "";
+}
+
+static void log(PlaceholderEntry* entry, const char* function, PlaceholderTable::classloadAction action) {
+  if (log_is_enabled(Debug, class, load, placeholders)) {
```

Comment:
Maybe `log()` should be explicitly declared as an inline function, so that the `if (log_is_enabled(Debug, class, load, placeholders))` can be checked without making a function call?

Reply:
OK.